Packages/com.unity.inputsystem/InputSystem/Actions/IInputInteraction.cs

```diff
@@ -19,17 +19,237 @@ namespace UnityEngine.InputSystem
-    /// this and also implement logic that signals cancellations (which the default logic never
-    /// triggers).
+    /// Actions have a built-in interaction pattern that to some extent depends on their type (<see
+    /// cref="InputActionType"/>, <see cref="InputAction.type"/>). What this means is that when controls
+    /// bound to an action are actuated, the action will initiate an interaction that in turn determines
+    /// when <see cref="InputAction.started"/>, <see cref="InputAction.performed"/>, and <see cref="InputAction.canceled"/>
+    /// are called.
+    ///
+    /// The default interaction (i.e. when no interaction has been added to a binding or the
+    /// action that the binding targets) will generally start and perform an action as soon as a control
+    /// is actuated, then perform the action whenever the value of the control changes except if the value
+    /// changes back to the default in which case the action is cancelled.
+    ///
+    /// By writing custom interactions, it is possible to implement different interactions. For example,
+    /// <see cref="Interactions.HoldInteraction"/> will only start when a control is being actuated but
+    /// will only perform the action if the control is held for a minimum amount of time.
     ///
     /// Interactions can be stateful and mutate state over time. In fact, interactions will usually
     /// represent miniature state machines driven directly by input.
+    ///
+    /// Multiple interactions can be applied to the same binding. The interactions will be processed in
+    /// sequence. However, the first interaction that starts the action will get to drive the state of
+    /// the action. If it performs the action, all interactions are reset. If it cancels, the first
+    /// interaction in the list that is in started state will get to take over and drive the action.
+    ///
+    /// This makes it possible to have several interaction patterns on the same action. For example,
+    /// to have a "fire" action that allows for charging, one can have a "Hold" and a "Press" interaction
+    /// in sequence on the action.
+    ///
+    /// <example>
+    /// <code>
+    /// // Create a fire action with two interactions:
+    /// // 1. Hold. Triggers charged firing. Has to come first as otherwise "Press" will immediately perform the action.
+    /// // 2. Press. Triggers instant firing.
+    /// // NOTE: An alternative is to use "Tap;Hold", i.e. a "Tap" first and then a "Hold". The difference
+    /// //       is relatively minor. In this setup, the "Tap" turns into a "Hold" if the button is held for
+    /// //       longer than the tap time whereas in the setup below, the "Hold" turns into a "Press" if the
+    /// //       button is released before the hold time has been reached.
+    /// var fireAction = new InputAction(type: InputActionType.Button, interactions: "Hold;Press");
```

Comment:
Is there a difference between "Hold;Press" and "Hold,Press"?

Reply:
Heh, didn't even know. Had to check. It's the same. For historical reasons, both the semicolon and comma are allowed in this case.

Reply:
You didn't know, but used both syntaxes in the same PR :) 

Do those historical reasons still apply? Should we allow only one for less confusion?

Reply:
>You didn't know, but used both syntaxes in the same PR :)

I did? Help me see it :)

>Do those historical reasons still apply? Should we allow only one for less confusion?

Dropping `;` is an option, I think. Initially, I used `,` everywhere as separators but for reasons that escape me now (think there was some data used as elements where `,` was already used) started switching over `;` with the intention of making that the element separator in all the JSON serialized data. However, `NameAndParameters` never really ended up fully making the switch and the editor is still writing out stuff like `Tap,Hold(duration=1,pressPoint=0.5)` instead of `Tap;Hold(duration=1,pressPoint=0.5)` (as it was intended to be).

So I guess we could drop `;` here.

Reply:
>Help me see it :)

CoreTests_Actions_Interactions.cs:17

I think dropping one of the syntaxes would be better. I wonder if it would be better to drop `,`, because `;` has the benefit of being distinguished from the `,` between parameters. But maybe that's not important?

Reply:
>CoreTests_Actions_Interactions.cs:17

Hah, indeed :)

> think dropping one of the syntaxes would be better. I wonder if it would be better to drop ,, because ; has the benefit of being distinguished from the , between parameters. But maybe that's not important?

I'd prefer that, too. Was one of the original motivations to change to the `,`. But problem is that change is actually a more breaking one. Then, I think, we'd have to somehow upgrade the .inputactions files that people have as those actually do use the `,` syntax instead of the `;` syntax. Which I'm thinking isn't worth the bother. Overall, allowing both `;` and `,` here seems like a relatively minor wart to me. Think I'm leaning towards "let's just live with it" side.

Reply:
Fair enough. Might try to be consistent in only using one in our docs then, though.